### PR TITLE
Orcasql 2018 06 01

### PR DESCRIPTION
**Pull request by hyunkyushim**
### Latest improvements:
<i>Port OrcaSQL APIs from 2018-06-01-privatepreview to 2018-06-01 (stable)
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/adx-documentation-pr/wiki/Overall-basic-flow) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PerformanceRecommendations.json [hunk 1]
```diff
@@ -0,0 +1,555 @@
...
+        }
+      }
+    },
+    "RecommendationActionProperties": {
+      "properties": {
+        "advisorName": {
+          "type": "string",
+          "description": "Advisor name."
+        },
+        "sessionId": {
```
Comment (pilor) [inline specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PerformanceRecommendations.json]: Please go through the models in all these files and mark properties that are "readOnly" as such with "readOnly": true.

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateEndpointConnections.json [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+            "$ref": "./examples/PrivateEndpointConnectionDelete.json"
+          }
+        }
+      },
+      "patch": {
+        "tags": [
+          "PrivateEndpointConnections"
+        ],
+        "operationId": "PrivateEndpointConnections_UpdateTags",
+        "summary": "Updates tags on private endpoint connection.",
```
Comment (pilor) [inline specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateEndpointConnections.json]: This uses a ProxyResource model below which would not normally have tags.  Is this actually a tracked resource?

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateEndpointConnections.json [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+        },
+        "type": {
+          "description": "Resource type.",
+          "type": "string",
+          "readOnly": true
+        }
+      },
+      "x-ms-azure-resource": true
+    },
+    "ProxyResource": {
```
Comment (pilor) [inline specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateEndpointConnections.json]: There is a common ProxyResource definition you can reference in specification\common-types\resource-management\v1\types.json

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateEndpointConnections.json [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+        },
+        "nextLink": {
+          "description": "Link to retrieve next page of results.",
+          "type": "string",
+          "readOnly": true
+        }
+      }
+    },
+    "CloudError": {
+      "x-ms-external": true,
```
Comment (pilor) [inline specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateEndpointConnections.json]: There is a common error definition you can reference in specification\common-types\resource-management\v1\types.json

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateEndpointConnections.json [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+            "type": "string"
+          },
+          "description": "Resource tags."
+        }
+      },
+      "description": "Tags object for patch operations."
+    }
+  },
+  "parameters": {
+    "SubscriptionIdParameter": {
```
Comment (pilor) [inline specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateEndpointConnections.json]: Please reference the common subscription, apiversion, and resourcegroup params from specification\common-types\resource-management\v1\types.json

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateLinkResources.json [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+      "authorizationUrl": "https://login.microsoftonline.com/common/oauth2/authorize",
+      "flow": "implicit",
+      "description": "Azure Active Directory OAuth2 Flow",
+      "scopes": {
+        "user_impersonation": "impersonate your user account"
+      }
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DBforMariaDB/servers/{serverName}/privateLinkResources": {
```
Comment (pilor) [inline specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateLinkResources.json]: Why is this "privateLinkResources"?  Its in Azure so by default its a "resource".  Should it be "privateLinks"?

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateLinkResources.json [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+        "operationId": "PrivateLinkResources_Get",
+        "parameters": [
+          {
+            "$ref": "#/parameters/ResourceGroupParameter"
+          },
+          {
+            "$ref": "#/parameters/ServerNameParameter"
+          },
+          {
+            "name": "groupName",
```
Comment (pilor) [inline specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateLinkResources.json]: The user is retrieving something called a "privateLinkResource" but being asked to provide a "groupName".  That doesn't make sense to me

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/QueryPerformanceInsights.json [hunk 8]
```diff
@@ -0,0 +1,684 @@
...
+            "$ref": "#/parameters/SubscriptionIdParameter"
+          },
+          {
+            "$ref": "#/parameters/ResourceGroupParameter"
+          },
+          {
+            "$ref": "#/parameters/ServerNameParameter"
+          },
+          {
+            "name": "parameters",
```
Comment (pilor) [inline specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/QueryPerformanceInsights.json]: GET APIs cannot have a body

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/QueryPerformanceInsights.json [hunk 9]
```diff
@@ -0,0 +1,684 @@
...
```
Comment (pilor) [inline specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/QueryPerformanceInsights.json]: GET cannot have a body

**Review by pilor — CHANGES_REQUESTED** [review]

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateEndpointConnections.json (hunk 3 above)
Comment (hyunkyushim) [reply specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateEndpointConnections.json]: Thanks. I was not aware. This is super helpful

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/QueryPerformanceInsights.json (hunk 9 above)
Comment (hyunkyushim) [reply specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/QueryPerformanceInsights.json]: So, these are parameters in the query. How should they be annotated?


**/topQueryStatistics** API
`
https://management.southcentralus1-c.control.orcasql-scus1-c.mscds.com/modules/ArmMySQL/subscriptions/ffffffff-ffff-ffff-ffff-ffffffffffff/resourceGroups/mysql_resource/providers/Microsoft.DBforMySQL/servers/test-server/topQueryStatistics?api-version=2018-06-01-privatepreview&numberOfTopQueries=5&aggregationFunction=avg&observedMetric=duration&observationStartTime=2019-04-01T20:00:00.000Z&observationEndTime=2019-05-13T20:15:00.000Z&aggregationWindow=PT15M
`
* numberOfTopQueries
* aggregationFunction
* observedMetric
* observationStartTime
* observationEndTime


**/waitStatistics** API
`
https://management.southcentralus1-c.control.orcasql-scus1-c.mscds.com/modules/ArmMySQL/subscriptions/ffffffff-ffff-ffff-ffff-ffffffffffff/resourceGroups/mysql_resource/providers/Microsoft.DBforMySQL/servers/test-server/waitStatistics?api-version=2018-06-01-privatepreview&observationStartTime=2019-05-05T23:30:00.000Z&observationEndTime=2019-05-06T23:30:00.000Z&aggregationWindow=PT15M
`
* observationStartTime
* observationEndTime
* aggregationWindow

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateEndpointConnections.json (hunk 2 above)
Comment (hyunkyushim) [reply specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateEndpointConnections.json]: tagging author @ramnov

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateLinkResources.json (hunk 7 above)
Comment (hyunkyushim) [reply specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateLinkResources.json]: tagging author @ramnov

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateLinkResources.json (hunk 6 above)
Comment (hyunkyushim) [reply specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateLinkResources.json]: tagging author @ramnov

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateEndpointConnections.json (hunk 4 above)
Comment (hyunkyushim) [reply specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateEndpointConnections.json]: tagging author @ramnov 
"ErrorResponse" definition in the mentioned types.json seems to be a superset of fields.

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/QueryPerformanceInsights.json (hunk 9 above)
Comment (pilor) [reply specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/QueryPerformanceInsights.json]: Are these results generated at API call time?  These seem more like query APIs than resource collection GETs.  How do the individual GETs work if the results are generated?

If it is generating results I would suggest using a POST action and either representing your parameters in [ODATA compliant query parameters](https://www.odata.org/documentation/odata-version-3-0/url-conventions/) if possible or as a request body if not.

**Comment by yungezz** [comment]
hi @hyunkyushim could you pls address comments from ARM reviewing?

**Comment by yungezz** [comment]
hi @hyunkyushim  pls address review comments.

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/QueryPerformanceInsights.json (hunk 9 above)
Comment (hyunkyushim) [reply specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/QueryPerformanceInsights.json]: These results are _not_ generated at API call time. They are continuously aggregated and persisted on the system tables as a background process while the feature is enabled. The APIs here retrieves the rows from the respective system table.

**Comment by sanjaiganesh** [comment]
@ramnov  COmments are not addressed yet. Please take a look.

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateEndpointConnections.json (hunk 2 above)
Comment (ramnov) [reply specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateEndpointConnections.json]: @sanjaiganesh suggested to add patch apis for atleast tags in this PR : https://github.com/Azure/azure-rest-api-specs-pr/pull/940

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateLinkResources.json (hunk 6 above)
Comment (ramnov) [reply specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateLinkResources.json]: This privateLinkResources name is dictated by the NRP team. This cannot be an arbitrary value.

File: specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateLinkResources.json (hunk 7 above)
Comment (ramnov) [reply specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateLinkResources.json]: Once again, this private link resources API is queried by Network Resource Provider and we have to adhere to their API specs.

Comment (ramnov) [reply specification/mariadb/resource-manager/Microsoft.DBforMariaDB/stable/2018-06-01/PrivateLinkResources.json]: you can learn more about group ids here : https://docs.microsoft.com/en-us/cli/azure/network/private-endpoint?view=azure-cli-latest#optional-parameters

**Comment by yungezz** [comment]
hi @hyunkyushim  could you pls address review comments? let us know when your update done and ready for another review. thanks.

**Comment by hyunkyushim** [comment]
@yungezz We've addressed all the comments since last ARM review

**Comment by yungezz** [comment]
hi @NullMDR  could you pls help to look at the sdk automation failure? It seems error from tool. thanks.

**Comment by NullMDR** [comment]
@yungezz LintDiff should be fixed first as autorest will also do lintdiff check

**Comment by yungezz** [comment]
HI @hyunkyushim could you pls fix lintdiff error https://dev.azure.com/azure-sdk/public/_build/results?buildId=229220&view=logs&j=688669d0-441c-57c3-cf6d-f89a22ccfa5d&t=923b5a90-e462-5994-c166-86509c6bf5f2 ? thanks.

**Comment by hyunkyushim** [comment]
@yungezz We've addressed the LintDiff issues

**Review by yungezz — APPROVED** [review]

**Pull request merged**